### PR TITLE
Add simple 2d shapes

### PR DESCRIPTION
**Pull request by PeanutbutterWarrior**
Added `arc()`, `circle()`, `ellipse()`, `point()`, `quad()`, `rect()`, `square()` and `triangle()`. Also adds `fill()` to do colors.
I was unable to make `rect()` support rounded corners, nothing will happen if it is given 5 or 8 arguments.
Doesn't support `ellipseMode()` or `rectMode()`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -127,7 +127,8 @@ dmypy.json
 # Pyre type checker
 .pyre/
 
+# VS Code
+.vscode
```
Comment (sbordeyne) [inline .gitignore]: needs a trailing /

File: pyprocessing/primitives_2d.py [hunk 2]
```diff
@@ -1,6 +1,57 @@
...
+    pp.windows.set_ellipse(x, y, width, height)
+
+
+def rect(x, y, width, height, *args):
+    global pp
+    if len(args) == 0:
+        pp.windows.set_rectangle(x, y, width, height)
+    elif len(args) == 1:
+        pp.windows.set_rounded_rectangle(x, y, width, height,
+                                         args[0], args[0], args[0], args[0])
```
Comment (sbordeyne) [inline pyprocessing/primitives_2d.py]: Styling. You should refactor this as 
```python
pp.windows.set_rounded_rectangle(
    x, y, width, height,
    args[0], args[0], args[0], args[0],
)
```

File: pyprocessing/primitives_2d.py [hunk 3]
```diff
@@ -1,6 +1,57 @@
...
+    elif len(args) == 4:
+        pp.windows.set_rounded_rectangle(x, y, width, height,
+                                         args[0], args[1], args[2], args[3])
```
Comment (sbordeyne) [inline pyprocessing/primitives_2d.py]: same as above, styling

File: pyprocessing/primitives_2d.py [hunk 4]
```diff
@@ -1,6 +1,57 @@
...
+    else:
+        raise TypeError
```
Comment (sbordeyne) [inline pyprocessing/primitives_2d.py]: When raising an Exception, always add a message for the developer to find what's wrong.

The else clause is unnecessary

I'd raise a ValueError over a TypeError here, but that can be discussed

File: pyprocessing/renderer/tkinter/window.py [hunk 5]
```diff
@@ -1,19 +1,22 @@
 from collections import deque
 import tkinter as tk
+from math import tau, degrees
```
Comment (sbordeyne) [inline pyprocessing/renderer/tkinter/window.py:4]: imports should be in alphabetical order, i.e. import math should be between import collections and import tkinter and degrees before tau

File: pyprocessing/renderer/tkinter/window.py [hunk 6]
```diff
@@ -1,19 +1,22 @@
...
 
 from pyprocessing.renderer.actions import Action
 
 
 class Window(tk.Frame):
     def __init__(self, master, pyprocessing):
+
```
Comment (sbordeyne) [inline pyprocessing/renderer/tkinter/window.py:11]: Unnecessary newline

File: pyprocessing/renderer/tkinter/window.py [hunk 7]
```diff
@@ -63,3 +67,57 @@ def set_line(self, x1, y1, x2, y2):
             x1, y1, x2, y2, fill=self.stroke.hex,
         )
         self.queued_actions.append(action)
+    
+    def set_ellipse(self, x, y, width, height):
+        x1 = x - (width +1)//2
```
Comment (sbordeyne) [inline pyprocessing/renderer/tkinter/window.py]: spaces between operators and operands

File: pyprocessing/renderer/tkinter/window.py [hunk 8]
```diff
@@ -63,3 +67,57 @@ def set_line(self, x1, y1, x2, y2):
...
+        x2 = x1 + width + 1
+        y2 = y1 + height + 1
+        action = Action(
+            self.canvas, 'create_rectangle',
+            x1, y1, x2, y2, fill=self.fill.hex, outline=self.stroke.hex
+        )
+        self.queued_actions.append(action)
+    
+    def set_rounded_rectangle(self, x, y, width, height, c1, c2, c3, c4):
+        # For when rect has 5 or 8 arguments
```
Comment (sbordeyne) [inline pyprocessing/renderer/tkinter/window.py:92]: Add a todo here

File: pyprocessing/renderer/tkinter/window.py [hunk 9]
```diff
@@ -63,3 +67,57 @@ def set_line(self, x1, y1, x2, y2):
...
+            x1, y1, x2, y2,
+            start=start, extent=angle, style=mode, fill=self.fill.hex, outline=self.stroke.hex
+        )
+        self.queued_actions.append(action)
+
+    def set_point(self, x, y):
+        action = Action(self.canvas, 'create_line',
+            x, y, x+1, y+1, fill=self.stroke.hex
+        )
+        self.queued_actions.append(action)
```
Comment (sbordeyne) [inline pyprocessing/renderer/tkinter/window.py]: missing newline at the end of the file

**Review by sbordeyne — CHANGES_REQUESTED** [review]

**Comment by sbordeyne** [comment]
When making changes, don't make new commits, but rebase your branch and amend the appropriate commits (when the change was introduced)

File: pyprocessing/renderer/tkinter/window.py (hunk 7 above)
Comment (sbordeyne) [reply pyprocessing/renderer/tkinter/window.py]: I didn't request a change for every single one of them, but the same comment is valid globally

**Comment by PeanutbutterWarrior** [comment]
Sorry, is that for making the requested changes or making changes to keep it up to date with the master branch? Am I rebasing Add-Simple-2d-Shapes onto master?

File: pyprocessing/primitives_2d.py (hunk 4 above)
Comment (PeanutbutterWarrior) [reply pyprocessing/primitives_2d.py]: What do you mean the else is unnecessary? Without it the error would always be raised, which you don't want, and there is no behaviour for different numbers of arguments, so an error would have to be raised.

Comment (sbordeyne) [reply pyprocessing/primitives_2d.py]: My bad, you're right, I thought you returned after every condition.

Comment (sbordeyne) [reply pyprocessing/primitives_2d.py]: The other comments still hold true though

**Comment by sbordeyne** [comment]
I meant for the requested changes.

It's so that each commit is good on its own, and so that the history doesn't look like a christmas tree. It also helps with git blame later on

**Comment by PeanutbutterWarrior** [comment]
I've made the changes. I think I did the rebase correctly.

File: examples/basicshapes.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+def draw():
+    line(50, 100, 150, 100)
+    circle(200, 100, 50)
+    ellipse(300, 100, 50, 25)
+    square(400, 100, 50)
+    rect(500, 100, 50, 25)
+    triangle(50, 200, 300, 250, 100, 300)
+    quad(400, 200, 500, 250, 350, 300, 375, 100)
+    arc(200, 400, 80, 80, 0, PI+QUARTER_PI, 'PIE')
```
Comment (sbordeyne) [inline examples/basicshapes.py]: space between operators

**Review by sbordeyne — CHANGES_REQUESTED** [review]
I missed one thing in the example, fix it and rebase onto the master branch (for the PVector change) and I'll merge it

File: examples/basicshapes.py (hunk 10 above)
Comment (PeanutbutterWarrior) [reply examples/basicshapes.py]: Sorry, missed that one

**Comment by PeanutbutterWarrior** [comment]
All fixed and rebased

**Pull request merged**